src/framed_write.rs

```diff
@@ -109,6 +109,9 @@ where
     }
     fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Result<(), Self::Error>> {
         let this = &mut *self;
+
+        ready!(Pin::new(&mut this.inner).poll_flush(cx).map_err(Into::into))?;
+
```

Comment:
I think this is wrong. We flush after every write we do, and if the inner flush would be Pending, it will wake up the task again until it returns Poll:Ready. There is no need to add an extra flush here when we have not sent any data.

Reply:
We *try* to `AsyncWrite::poll_flush()` after every write, but this flush might return `Pending`. During the next call to `FramedWrite::poll_flush()`, we may need to flush again in order to make space in the I/O.

This really depends on whether or not `AsyncWrite::poll_write()` can make progress towards actually sending the data to its destination. The API [docs](https://docs.rs/futures-preview/0.3.0-alpha.17/futures/io/trait.AsyncWrite.html#tymethod.poll_write) don't say, one way or the other, whether or not a flush is *required* in order to make progress. They do say that poll_write

> arranges for the current task (via cx.waker().wake_by_ref()) to receive a notification when the object becomes writable or is closed

So perhaps this means that `AsyncWrite::poll_write()` makes progress on its own, and my flushes are superfluous.

@Nemo157, do you have any thoughts on this? Are flushes strictly necessary in order for the I/O to make progress towards its send/write?

Reply:
Based on what I said on the top-level comment both this flush and the one inside the `while` loop below is unnecessary, calling `AsyncWrite::poll_write` will implicitly flush data out and use whatever space becomes available.

Instead, there should be a `poll_flush` call _after_ the while loop, so that once you complete writing all the data to the underlying writer you wait till it has finished flushing it out before returning `Ready`.

Reply:
Removed spurious flushes in edc7a1f675.